pallets/migrations/src/lib.rs

```diff
@@ -0,0 +1,278 @@
+			// already in progress)
+			//		notably, we might have started an individual migration and the list of
+			//		migrations might change on our next on_runtime_upgrade()
+			<FullyUpgraded<T>>::put(false);
+			weight += T::DbWeight::get().writes(1);
+			Self::deposit_event(Event::RuntimeUpgradeStarted());
+
+			weight += process_runtime_upgrades::<T>(available_weight - weight);
+
+			// at least emit a warning if we aren't going to end up finishing our migrations...
+			if !T::MultiBlockMigrationsSupported::get() {
+				if !<FullyUpgraded<T>>::get() {
+					log::warn!(
+						"migrations weren't completed in on_runtime_upgrade(), but we're not
+					configured for multi-block migrations; state is potentially inconsistent!"
+					);
+				}
+			}
+
+			weight
+		}
+
+		/// on_initialize implementation. Calls process_runtime_upgrades() if we are still in the
+		/// middle of a runtime upgrade.
+		fn on_initialize(_: T::BlockNumber) -> Weight {
+			let mut weight: Weight = T::DbWeight::get().reads(1 as Weight);
+
+			// TODO: derive a suitable value here, which is probably something < max_block
+			let available_weight: Weight = T::BlockWeights::get().max_block;
+
+			if T::MultiBlockMigrationsSupported::get() {
+				if !<FullyUpgraded<T>>::get() {
+					weight += process_runtime_upgrades::<T>(available_weight);
+				}
+			}
+
+			weight
+		}
+	}
+
+	#[pallet::storage]
+	#[pallet::getter(fn is_fully_upgraded)]
+	/// True if all required migrations have completed
+	type FullyUpgraded<T: Config> = StorageValue<_, bool, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn migration_state)]
+	/// MigrationState tracks the progress of a migration.
+	/// Maps name (Vec<u8>) -> migration progress (Perbill)
+	type MigrationState<T: Config> = StorageMap<_, Twox64Concat, Vec<u8>, Perbill, OptionQuery>;
+
+	#[pallet::genesis_config]
+	pub struct GenesisConfig<T: Config> {
+		pub completed_migrations: Vec<Vec<u8>>,
+		pub phantom: PhantomData<T>,
+	}
+
+	#[cfg(feature = "std")]
+	impl<T: Config> Default for GenesisConfig<T> {
+		fn default() -> Self {
+			Self {
+				completed_migrations: vec![],
+				phantom: PhantomData,
+			}
+		}
+	}
+
+	#[pallet::genesis_build]
+	impl<T: Config> GenesisBuild<T> for GenesisConfig<T> {
+		fn build(&self) {
+			for migration_name in &self.completed_migrations {
+				<MigrationState<T>>::insert(migration_name, Perbill::one());
+			}
+		}
+	}
+
+	fn process_runtime_upgrades<T: Config>(available_weight: Weight) -> Weight {
+		log::info!("stepping runtime upgrade");
+
+		let mut weight: Weight = 0u64.into();
+		let mut done: bool = true;
+
+		for migration in &T::MigrationsList::get() {
+			// let migration_name = migration.friendly_name();
+			let migration_name = migration.friendly_name();
+			let migration_name_as_bytes = migration_name.as_bytes();
+			log::trace!("evaluating migration {}", migration_name);
+
+			let migration_state =
+				<MigrationState<T>>::get(migration_name_as_bytes).unwrap_or(Perbill::zero());
```

Comment:
I think I would not use the string as identifier.
It's somewhat unlikely, but the keys could collide

Reply:
What would you suggest?

Reply:
mmh, maybe it would work if you scoped it by pallet? (so append the name from pallet_info before the migration)